### PR TITLE
Add ImageMagick 7 checks to rmimage.c

### PR DESCRIPTION
**Pull request by dlemstra**

### CODE REVIEW

File: ext/RMagick/rmimage.c [hunk 1]
```diff
@@ -144,8 +148,11 @@ adaptive_method(int argc, VALUE *argv, VALUE self
  * @return a new image
  */
 static VALUE
-adaptive_channel_method(int argc, VALUE *argv, VALUE self
-                        , Image *fp(const Image *, const ChannelType, const double, const double, ExceptionInfo *))
+#if defined(IMAGEMAGICK_7)
+adaptive_channel_method(int argc, VALUE *argv, VALUE self, Image *fp(const Image *, const double, const double, ExceptionInfo *))
```
Comment (Watson1978) [inline ext/RMagick/rmimage.c]: It is better to use `#if ~ #else ~ # endif` only where it is needed.

```c
#if defined(IMAGEMAGICK_7)
    typedef Image *(channel_method_t)(const Image *, const double, const double, ExceptionInfo *);
#else
    typedef Image *(channel_method_t)(const Image *, const ChannelType, const double, const double, ExceptionInfo *);
#endif

...

static VALUE
adaptive_channel_method(int argc, VALUE *argv, VALUE self, channel_method_t fp)
{
...
```

File: ext/RMagick/rmimage.c [hunk 2]
```diff
@@ -861,10 +935,17 @@ crisscross(int bang, VALUE self, Image *fp(const Image *, ExceptionInfo *))
  * @return a new image
  */
 static VALUE
+#if defined(IMAGEMAGICK_7)
+auto_channel(int argc, VALUE *argv, VALUE self, MagickBooleanType (*fp)(Image *, ExceptionInfo *exception))
```
Comment (Watson1978) [inline ext/RMagick/rmimage.c]: Like `adaptive_channel_method`, it is better to use #if ~ #else ~ # endif only where it is needed.

File: ext/RMagick/rmimage.c [hunk 3]
```diff
@@ -1134,9 +1233,27 @@ VALUE
...
     image = rm_check_destroyed(self);
-    return rb_float_new(image->bias);
+#if defined(IMAGEMAGICK_7)
+    {
+        const char *artifact = GetImageArtifact(image, "convolve:bias");
+        if (artifact != (const char *) NULL)
+        {
+            char *q;
+
+            bias=InterpretLocaleValue(artifact,&q);
```
Comment (Watson1978) [inline ext/RMagick/rmimage.c]: It is better to have the space around `=`.

File: ext/RMagick/rmimage.c [hunk 4]
```diff
@@ -1134,9 +1233,27 @@ VALUE
...
+            if (*q == '%')
+            {
+                bias*=((double) QuantumRange+1.0)/100.0;
```
Comment (Watson1978) [inline ext/RMagick/rmimage.c]: It is better to have the space around `*=`.

Comment (Watson1978) [inline ext/RMagick/rmimage.c]:
```suggestion
                bias *= ((double) QuantumRange + 1.0) / 100.0;
```

File: ext/RMagick/rmimage.c [hunk 5]
```diff
@@ -3609,16 +3923,31 @@ composite_tiled(int bang, int argc, VALUE *argv, VALUE self)
...
         for (x = 0; status == MagickTrue && x < (long) image->columns; x += columns)
         {
+#if defined(IMAGEMAGICK_7)
+            BEGIN_CHANNEL_MASK(image, channels);
+            status = CompositeImage(image, comp_image, MagickTrue, operator, x, y, exception);
+            END_CHANNEL_MASK(image);
+            rm_check_exception(exception, image, bang ? RetainOnError: DestroyOnError);
+#else
             status = CompositeImageChannel(image, channels, operator, comp_image, x, y);
             rm_check_image_exception(image, bang ? RetainOnError: DestroyOnError);
```
Comment (Watson1978) [inline ext/RMagick/rmimage.c]: if exception was raised in `rm_check_image_exception()`, it can not release a heap area of `exception`.

File: ext/RMagick/rmimage.c [hunk 6]
```diff
@@ -4104,14 +4484,68 @@ Image_morphology_channel(VALUE self, VALUE channel_v, VALUE method_v, VALUE iter
...
+#endif
     rm_check_exception(exception, new_image, DestroyOnError);
     (void) DestroyExceptionInfo(exception);
 
     rm_ensure_result(new_image);
     return rm_image_new(new_image);
 }
 
+#if defined(IMAGEMAGICK_7)
+// TODO: Move this to KernelInfo class as a constructor?
```
Comment (Watson1978) [inline ext/RMagick/rmimage.c]: Need to fix TODO ?

File: ext/RMagick/rmimage.c [hunk 7]
```diff
@@ -7419,10 +8100,17 @@ Image_import_pixels(int argc, VALUE *argv, VALUE self)
 
     if (!okay)
     {
+#if defined(IMAGEMAGICK_7)
+        CHECK_EXCEPTION()
```
Comment (Watson1978) [inline ext/RMagick/rmimage.c]: In below, `rm_magick_error()` will raise an exception. So it needs to destroy `exception` to avoid memory leak.

File: ext/RMagick/rmimage.c [hunk 8]
```diff
@@ -9787,29 +10776,45 @@ Image_pixel_color(int argc, VALUE *argv, VALUE self)
...
     {
+#if defined(IMAGEMAGICK_7)
+        okay = SetImageStorageClass(image, DirectClass, exception);
+#else
         okay = SetImageStorageClass(image, DirectClass);
         rm_check_image_exception(image, RetainOnError);
+#endif
         if (!okay)
         {
             rb_raise(Class_ImageMagickError, "SetImageStorageClass failed. Can't set pixel color.");
```
Comment (Watson1978) [inline ext/RMagick/rmimage.c]: Before this line, it needs to destroy `exception` to avoid memory leak.

File: ext/RMagick/rmimage.c (hunk 1 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Good idea

File: ext/RMagick/rmimage.c (hunk 5 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: But `exception` was created inside an `#if defined(IMAGEMAGICK_7)` block?

File: ext/RMagick/rmimage.c (hunk 6 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: I added this as a TODO because I think this should be done after IM7 support has been added. I can remove the TODO if you want me to.

File: ext/RMagick/rmimage.c (hunk 8 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: `exception` will only be created inside an `#if defined(IMAGEMAGICK_7)` block?

**Comment by dlemstra** [comment]
Thanks for all the review comments @Watson1978!

File: ext/RMagick/rmimage.c (hunk 7 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Good catch!

**Comment by mockdeep** [comment]
@Watson1978 mind having another look?

**Comment by Watson1978** [comment]
By code reading, basically I don't know this change will generate same images
It is difficult to approve until we will add appearance test.

**Comment by dlemstra** [comment]
The pull requests called "Changes for ImageMagick 7" are all commits from my IM7 pull request. That PR also checks if all the unit tests are passing and all of them are passing. I don't think we need to wait with merging these changes before we have appearance tests in place. Your thoughts @mockdeep?

**Comment by mockdeep** [comment]
@Watson1978 can you give an example of what sort of test you'd like to see?

**Comment by Watson1978** [comment]
@mockdeep It is tough to secure time because my wife got an injury that required rehabilitation for now.

**Comment by mockdeep** [comment]
@Watson1978 no pressure, take care of your family first.

I don't think these PRs need to wait for end-to-end tests, but maybe we should make a point of having them before releasing IM7 support. @dlemstra would you be up for separately adding some tests like the montage one linked? It would probably be prohibitive to shoot for 100% feature coverage, but maybe we can touch on some of the more commonly used stuff.

Side note, I've been changing the commits when I've been merging to something a little more specific, like  `Add ImageMagick 7 checks to rmilist.c (#776)`. Just something that allows us to trace back to the commit for a specific file.

**Comment by dlemstra** [comment]
I don't think that I will have that much time for adding those tests. My Magick.NET library has around 2200 unit tests and a lot of them do visual testing so I think a lot has been covered there already. That does not mean we should not add those tests here but I really don't think that we need to wait with IM7 support until we have them. I would prefer to have a preview release so we can already have some field testing and get feedback from there.

**Comment by mockdeep** [comment]
Okay, that sounds alright to me. I feel a little nervous about merging these without @Watson1978, but I suppose he can come back and help us fix our mistakes when things calm down for him.

File: ext/RMagick/rmimage.c (hunk 6 above)
Comment (mockdeep) [reply ext/RMagick/rmimage.c]: Should we maybe open an issue for this instead?

File: ext/RMagick/rmimage.c [hunk 9]
```diff
@@ -2307,7 +2503,13 @@ Image_channel_entropy(int argc, VALUE *argv, VALUE self)
     }
 
     exception = AcquireExceptionInfo();
+#if IMAGEMAGICK_7
+    BEGIN_CHANNEL_MASK(image, channels);
+    (void) GetImageEntropy(image, &entropy, exception);
+    END_CHANNEL_MASK(image);
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: I would not prefer that. At some point we will remove the IM6 code and then we would need to move those changes into rmimage.c I would really prefer to keep it this way.

File: ext/RMagick/rmimage.c [hunk 10]
```diff
@@ -3296,17 +3575,35 @@ composite(int bang, int argc, VALUE *argv, VALUE self, ChannelType channels)
 
     if (bang)
     {
+#if defined(IMAGEMAGICK_7)
+        exception = AcquireExceptionInfo();
+        BEGIN_CHANNEL_MASK(image, channels);
+        (void) CompositeImage(image, comp_image, operator, MagickTrue, x_offset, y_offset, exception);
+        END_CHANNEL_MASK(image);
+        CHECK_EXCEPTION()
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: In this spot the image is the input image and if we would call `rm_check_exception` here instead and an exception was raised the image would be destroyed.

File: ext/RMagick/rmimage.c [hunk 11]
```diff
@@ -3296,17 +3575,35 @@ composite(int bang, int argc, VALUE *argv, VALUE self, ChannelType channels)
...
     else
     {
         new_image = rm_clone_image(image);
 
+#if defined(IMAGEMAGICK_7)
+        exception = AcquireExceptionInfo();
+        BEGIN_CHANNEL_MASK(new_image, channels);
+        (void) CompositeImage(new_image, comp_image, operator, MagickTrue, x_offset, y_offset, exception);
+        END_CHANNEL_MASK(new_image);
+        rm_check_exception(exception, new_image, DestroyOnError);
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Here we called `new_image = rm_clone_image(image);` to create a new image and this should be destroyed when an exception is raised.

File: ext/RMagick/rmimage.c [hunk 12]
```diff
@@ -3720,10 +4045,21 @@ Image_compress_colormap_bang(VALUE self)
...
     Image *image;
     MagickBooleanType okay;
+#if defined(IMAGEMAGICK_7)
+    ExceptionInfo *exception;
+#endif
 
     image = rm_check_frozen(self);
+
+#if defined(IMAGEMAGICK_7)
+    exception = AcquireExceptionInfo();
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Most of the RMagick methods have an `rm_` prefix and there is not a real convention for the ImageMagick methods.

File: ext/RMagick/rmimage.c [hunk 13]
```diff
@@ -3854,27 +4190,45 @@ Image_constitute(VALUE class ATTRIBUTE_UNUSED, VALUE width_arg, VALUE height_arg
         rb_raise(rb_eNoMemError, "not enough memory to continue.");
     }
 
+#if defined(IMAGEMAGICK_7)
+    exception = AcquireExceptionInfo();
+    SetImageExtent(new_image, width, height, exception);
+#else
     SetImageExtent(new_image, width, height);
     exception = &new_image->exception;
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: These changes where made to prevent leaking of pixels.v. But it does look like you found a mistake here. Very good catch. For IM6 the exception will be freed and will result in an access violation. Will push a patch to fix this.

File: ext/RMagick/rmimage.c [hunk 14]
```diff
@@ -4104,14 +4480,68 @@ Image_morphology_channel(VALUE self, VALUE channel_v, VALUE method_v, VALUE iter
 
     exception = AcquireExceptionInfo();
 
+#if defined(IMAGEMAGICK_7)
+    BEGIN_CHANNEL_MASK(image, channel);
+    new_image = MorphologyImage(image, method, NUM2LONG(iterations), kernel, exception);
+    CHANGE_RESULT_CHANNEL_MASK(new_image);
+    END_CHANNEL_MASK(image);
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: That would not really be an option here because the signature of the methods inside that block would be very different and you would need a method for every signature.

File: ext/RMagick/rmimage.c [hunk 15]
```diff
@@ -4388,7 +4859,11 @@ Image_density(VALUE self)
 
     image = rm_check_destroyed(self);
 
+#if defined(IMAGEMAGICK_7)
+    sprintf(density, "%gx%g", image->resolution.x, image->resolution.y);
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: We could do that with a define but at some point the IM6 code would be remove and then we still need to change it. Not a big fan of using an alias here but I will let you decide?

File: ext/RMagick/rmimage.c [hunk 16]
```diff
@@ -4571,12 +5059,21 @@ VALUE
...
+#endif
 
     image = rm_check_frozen(self);
 
-    // Store a clone of the mask image
+#if defined(IMAGEMAGICK_7)
+    exception = AcquireExceptionInfo();
+    (void) SetImageMask(image, CompositePixelMask, NULL, exception);
+    CHECK_EXCEPTION()
+    (void) DestroyExceptionInfo(exception);
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: You would need a method (with a different name) for each signature so I don't think that would feasible here. And I don't think it would make the code very readable.

p.s. After IM7 has been merged I would really like to get rid of the `CHECK_EXCEPTION` defined because it is not clear what it does when you read the code like this. But that will be a discusion for another moment.

File: ext/RMagick/rmimage.c [hunk 17]
```diff
@@ -6277,7 +6829,11 @@ Image_find_similar_region(int argc, VALUE *argv, VALUE self)
     }
 
     exception = AcquireExceptionInfo();
+#if defined(IMAGEMAGICK_7)
+    okay = IsEquivalentImage(image, target, &x, &y, exception);
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: They do. This method has been renamed inside ImageMagick 7.

File: ext/RMagick/rmimage.c [hunk 18]
```diff
@@ -7143,6 +7784,30 @@ has_attribute(VALUE self, MagickBooleanType (attr_test)(const Image *, Exception
...
+ * Run a function testing whether this image has an attribute.
+ *
+ * No Ruby usage (internal function)
+ *
+ * @param self this object
+ * @param attr_test the attribute testing function
+ * @return the result of attr_test.
+ */
+static VALUE
+has_image_attribute(VALUE self, MagickBooleanType (attr_test)(const Image *))
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: The `has_attribute` method calls a method that has an `image` and `exception` argument and `has_image_attribute` calls a method that only needs an `image`.

File: ext/RMagick/rmimage.c [hunk 19]
```diff
@@ -8409,6 +9153,82 @@ set_image_mask(Image *image, VALUE mask)
...
+ * No Ruby usage (internal function)
+ *
+ * @param image the image
+ * @param mask the mask
+ * @param type the mask type
+ * @return copy of the current clip-mask or nil
+ * @see get_image_mask
+ */
+static VALUE
+set_image_mask(Image *image, VALUE mask, PixelMask type)
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: We could probably default to `WritePixelMask` for now but in the future we could add support for other masks. I would prefer to keep the argument for future use but I will let you decide?

File: ext/RMagick/rmimage.c [hunk 20]
```diff
@@ -9504,7 +10456,11 @@ Image_ordered_dither(int argc, VALUE *argv, VALUE self)
 
     exception = AcquireExceptionInfo();
 
+#if (IMAGEMAGICK_7)
+    (void) OrderedDitherImage(new_image, threshold_map, exception);
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: `OrderedPosterizeImage` actually does an ordered dither and that is also why the ruby method has this name:

```
%  OrderedPosterizeImage() will perform a ordered dither based on a number
%  of pre-defined dithering threshold maps, but over multiple intensity
%  levels, which can be different for different channels, according to the
%  input argument.
```

We have corrected the method name in IM7.

File: ext/RMagick/rmimage.c [hunk 21]
```diff
@@ -9762,11 +10740,18 @@ Image_pixel_color(int argc, VALUE *argv, VALUE self)
...
+        old_pixel = GetVirtualPixels(image, x, y, 1, 1, exception);
         CHECK_EXCEPTION()
 
         (void) DestroyExceptionInfo(exception);
 
+#if defined(IMAGEMAGICK_7)
+        old_color.red   = GetPixelRed(image, old_pixel);
+        old_color.green = GetPixelGreen(image, old_pixel);
+        old_color.blue  = GetPixelBlue(image, old_pixel);
+        old_color.alpha = GetPixelAlpha(image, old_pixel);
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: For `IM6` there is already a define called `GetPixelRed` but we cannot use that here because it has it does not have the `image` argument.

File: ext/RMagick/rmimage.c [hunk 22]
```diff
@@ -10279,11 +11307,17 @@ Image_quantize(int argc, VALUE *argv, VALUE self)
             if (rb_obj_is_kind_of(argv[2], Class_DitherMethod))
             {
                 VALUE_TO_ENUM(argv[2], quantize_info.dither_method, DitherMethod);
+#if !defined(IMAGEMAGICK_7)
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: It would be weird to do that and for the the `!` is pretty normal but I can add an `IMAGEMAGICK_6` define if you would prefer that. If you really want that could I do that in a new PR because it will also touch other files. You decide?

File: ext/RMagick/rmimage.c [hunk 23]
```diff
@@ -10441,7 +11487,18 @@ Image_random_threshold_channel(int argc, VALUE *argv, VALUE self)
 
     exception = AcquireExceptionInfo();
 
+#if defined(IMAGEMAGICK_7)
+    BEGIN_CHANNEL_MASK(new_image, channels);
+    {
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: They are there to make sure that `GeometryInfo geometry_info;` is defined in a separate block and this is the style we use inside ImageMagick. And added the block because we really only need this for IM7.

File: ext/RMagick/rmimage.c (hunk 6 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Lets open it once this has merged?

File: ext/RMagick/rmimage.c (hunk 8 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Aaah here was the comment :) Edited your message mockdeep and this is indeed a memory leak thanks for finding this @Watson1978.

**Comment by dlemstra** [comment]
Thanks for taking the time to review this @mockdeep! Feel free to contact me if you want to talk about some of these remarks in a video chat. Hope we can merge this soon 💃

File: ext/RMagick/rmimage.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: I have also thought a lot about this and this was the most maintainable solution that I could come up with. There is a big difference in the method signatures because most IM7 methods now don't have a channel version of the method and have an exceptioninfo argument. But the setup and checks in RMagick are the same for most of these calls. Most changes have been isolated to how the ImageMagick API is being called.

File: ext/RMagick/rmimage.c (hunk 22 above)
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Okay will do that in a new PR after this one is merged.

File: ext/RMagick/rmimage.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: We will not make any changes to the API of ImageMagick 6. The API for IM6 is frozen and we will only do bug fixes.

File: ext/RMagick/rmimage.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Yeah that sounds like a good option. Just pushed a patch for this.

File: ext/RMagick/rmimage.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Done!

File: ext/RMagick/rmimage.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: My comment got lost. These braces are here because `GeometryInfo geometry_info;` is not used by the other method. But it is probably better to move that up and in a `if defined(IMAGEMAGICK_7)` block. I will let you decide?

[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: I placed it in the braces so the variables are not declared in the middle of the code but at the start of a scope (code within braces). I could move the declartion up near `ExceptionInfo *exception;` and remove the braces but I would need another `if defined(IMAGEMAGICK_7)` block there then. You decide?

File: ext/RMagick/rmimage.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: `image->resolution.x, , image->resolution.x, image->resolution.y` would become something like `RESOLUTION_X(image), RESOLUTION_Y(image)`. This does not really improve the readability. But you decide?

File: ext/RMagick/rmimage.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Because the image contains information at which index in the pointer the specified channel is. If the image is for example `RGB` the index for `R=0`, `G=1` and `B=2`. And that needs to be used as the offset from the `old_pixel` pointer. `B` would be `*(pointer+2)`. And if the order would be `GBR` the index would become `1` for `B`. For ImageMagick 6 the order is fixed instead.

File: ext/RMagick/rmimage.c [hunk 24]
```diff
@@ -2307,7 +2503,13 @@ Image_channel_entropy(int argc, VALUE *argv, VALUE self)
     }
 
     exception = AcquireExceptionInfo();
+#if IMAGEMAGICK_7
```
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Comment is gone here also... Good catch this has been resolved.

File: ext/RMagick/rmimage.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (mockdeep) [reply ext/RMagick/rmimage.c]: Hmm, yeah, that's not ideal. I was hoping there might be a way to monkey patch `image` to have a `resolution` property against IM6 so that we could call `image->resolution.x` for both cases, but it doesn't sound like it. I guess let's leave it like this for now. I may revisit as I get a better understanding of C.

File: ext/RMagick/rmimage.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (mockdeep) [reply ext/RMagick/rmimage.c]: I see. I think it's fine to leave it like this.

File: ext/RMagick/rmimage.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (mockdeep) [reply ext/RMagick/rmimage.c]: I see. Thanks for the explanation!

File: ext/RMagick/rmimage.c [hunk 25]
```diff
@@ -2289,7 +2486,7 @@ Image_channel_mean(int argc, VALUE *argv, VALUE self)
 VALUE
 Image_channel_entropy(int argc, VALUE *argv, VALUE self)
 {
-#if defined(HAVE_GETIMAGECHANNELENTROPY)
+#if defined(HAVE_GETIMAGECHANNELENTROPY) || defined(IMAGEMAGICK_7)
```
Comment (mockdeep) [inline ext/RMagick/rmimage.c]: I wonder if it would be worth using `IMAGEMAGICK_GREATER_THAN_EQUAL_6_9_0` for these cases.

File: ext/RMagick/rmimage.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (mockdeep) [reply ext/RMagick/rmimage.c]: I agree with removing `CHECK_EXCEPTION`. I've been reading it like a function call when it is a define, and it doesn't really do a whole lot for us. In general, I'm beginning to feel like `#define` should be avoided when possible, unless it provides a *lot* of value.

File: ext/RMagick/rmimage.c [hunk 26]
```diff
@@ -9787,29 +10738,51 @@ Image_pixel_color(int argc, VALUE *argv, VALUE self)
...
+        okay = SetImageStorageClass(image, DirectClass, exception);
+        CHECK_EXCEPTION()
+#else
         okay = SetImageStorageClass(image, DirectClass);
-        rm_check_image_exception(image, RetainOnError);
+        if (rm_should_raise_exception(&image->exception, RetainExceptionRetention))
+        {
+            (void) DestroyExceptionInfo(exception);
+            rm_check_image_exception(image, RetainOnError);
+        }
```
Comment (mockdeep) [inline ext/RMagick/rmimage.c]: This is unrelated to the IM7 upgrade, right? Maybe it would be worth pulling it out into a new PR.

File: ext/RMagick/rmimage.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (mockdeep) [reply ext/RMagick/rmimage.c]: I'm fine with moving it up, too, if you prefer.

File: ext/RMagick/rmimage.c [hunk 27]
```diff
@@ -12796,36 +14038,48 @@ Image_store_pixels(VALUE self, VALUE x_arg, VALUE y_arg, VALUE cols_arg
...
+#if defined(IMAGEMAGICK_7)
+    okay = SetImageStorageClass(image, DirectClass, exception);
+    CHECK_EXCEPTION()
+#else
     okay = SetImageStorageClass(image, DirectClass);
     rm_check_image_exception(image, RetainOnError);
+#endif
     if (!okay)
     {
+        (void) DestroyExceptionInfo(exception);
```
Comment (mockdeep) [inline ext/RMagick/rmimage.c]: This looks like maybe another fix for an existing bug that should be pulled out into another PR.

**Comment by mockdeep** [comment]
Just a few more comments @dlemstra, but this is looking really good. You've put a monumental amount of work into this, so it is very appreciated.

File: ext/RMagick/rmimage.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Done!

File: ext/RMagick/rmimage.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Will just leave it like this.

File: ext/RMagick/rmimage.c (hunk 25 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Okay if we move this to a new PR? And maybe we could also replace `HAVE_SETIMAGEGRAY` with`IMAGEMAGICK_GREATER_THAN_EQUAL_6_9_2` there?

File: ext/RMagick/rmimage.c (hunk 26 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: This is related to the ImageMagick 7 upgrade because `exception = AcquireExceptionInfo();` has been moved up for IM7.

File: ext/RMagick/rmimage.c (hunk 27 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Not a fix for an existing bug I just pushed a patch to make it more readable

File: ext/RMagick/rmimage.c (hunk 26 above)
Comment (dlemstra) [reply ext/RMagick/rmimage.c]: Just pushed another path to make this more readable.

File: ext/RMagick/rmimage.c (hunk 25 above)
Comment (mockdeep) [reply ext/RMagick/rmimage.c]: Sure, sounds good.

**Comment by mockdeep** [comment]
Alright, bombs away!
